src/qtpromise/qpromisehelpers_p.h

```diff
@@ -88,6 +88,61 @@ void connectDestroyedToReject(const QtPromise::QPromiseConnections& connections,
     });
 }
 
+template<typename T, typename U>
+struct AreArithmeticTypes
+    : std::integral_constant<bool, std::is_arithmetic<T>::value && std::is_arithmetic<U>::value>
+{ };
+
+template<typename From, typename To, bool = AreArithmeticTypes<From, To>::value>
+struct PromiseConverter;
```

Comment:
I'd move the declarations to qpromise_p.h and the implementations to qpromise.inl since the QVariant->U convertor throws an exception defined in qpromiseexceptions.h. Is this okay or should there be a `qpromise_p.inl`?

Reply:
Having `PromiseConverter` in a new file will be inconsistent with other internal classes. And at some point, I'm thinking to merge `qpromise.inl` in `qpromise.h` to be consistent with Qt. So I'm a bit reluctant to introduce a new `qpromise_p.inl` file.

I think logic in `qpromise_p.h` should be able to throw QPromise* exceptions. I'm wondering if moving `QPromiseError` (deprecated) at the end of `qpromise_p.h`  (so we can remove the `qpromise_p.h` include) would be a breaking change (if someone included `qpromiseexceptions.h` directly)?

Reply:
I've started implementing your feedback today, and the problem is not only in QPromise* exceptions but it is actually the implementations of the ` PromiseConverter::call()` themselves: they require the return type `QPromise<T>`  to be complete at this point. So I've ended up puttint the ` PromiseConverter`  implementations at the end of ` qpromise.h` so far. 

Regarding `QPromiseError`: is `qpromiseexceptions.h` supposed to be included directly? I think there could be a preprocessor check in `qpromiseexceptions.h` to verify whether `qpromise_p.h` has been included before qpromiseexceptions.h or not. If not, then we know that somebody has included `qpromsieexceptions.h`  directly, and we could emit a compiler error saying to update the includes, something like:

```c++
// qpromisexceptions.h

#if !defined(QTPROMISE_QPROMISE_P_H)
#error Do not use qpromiseexceptions.h, include QtPromise instead.
#endif
```

It's not really elegant, but it should work

Reply:
> ...is qpromiseexceptions.h supposed to be included directly?

Yes, all files that don't have the `_p.h` suffix can be included directly. Though, as written in the comment, `QPromiseError` was not documented and wasn't supposed to be used outside the lib and since it's unlikely someone included `qpromiseexceptions.h` directly,I would not try to make this case non breaking. I will mention it in the release notes.

> ...they require the return type QPromise<T> to be complete at this point.

Could we return the lambda instead of passing the promise to the converter:

```c
template<typename T, typename U>
struct PromiseConverter
{
    // NOTE: can we use `auto` instead of `std::function`?
    static std::function<U(const T&)> create()
    {
        return [](const T& value) {
            return static_cast<U>(value);
        };
    }
};

// ...

template<typename T>
struct PromiseConverter<T, void>
{
    static std::function<void()> create()
    {
        return []() {};
    };
};

// ...

template<typename T>
template<typename U>
inline QPromise<U> QPromiseBase<T>::as() const
{
  return then(QtPromisePrivate::PromiseConverter<T, U>::create());
}
```

Reply:
> Could we return the lambda instead 
Oh wow, that's a neat trick! Worked like a charm! I've been able to put the converters into `qpromise_p.h` now, thanks